crates/cli/cli/src/commands/query/client/consensus.rs

```diff
@@ -0,0 +1,110 @@
+    )]
+    client_id: ClientId,
+
+    #[clap(
+        long = "consensus-height",
+        value_name = "CONSENSUS_HEIGHT",
+        help = "Height of the client's consensus state to query, if not specified all consensus heights are returned"
+    )]
+    consensus_height: Option<u64>,
+
+    #[clap(
+        long = "height",
+        value_name = "HEIGHT",
+        help = "The height at which to query the chain"
+    )]
+    height: Option<u64>,
+}
+
+impl CommandRunner<CosmosBuilder> for QueryClientConsensus {
+    async fn run(&self, builder: &CosmosBuilder) -> Result<Output> {
+        let chain = builder.build_chain(&self.chain_id).await?;
+
+        let counterparty_chain_id = <CosmosChain as CanQueryClientStateWithLatestHeight<
+            CosmosChain,
+        >>::query_client_state_with_latest_height(
+            &chain, &self.client_id
+        )
+        .await
+        .map(|cs| cs.chain_id)
+        .wrap_err("failed to query counterparty chain from client state")?;
+
+        if let Some(consensus_height) = self.consensus_height {
+            let consensus_height = Height::new(counterparty_chain_id.version(), consensus_height)
+                .wrap_err_with(|| format!(
+                    "failed to create Height with revision number `{}` and revision height `{consensus_height}`",
+                    counterparty_chain_id.version()
+                ))?;
+
+            let query_height = self.height.map(|height| {
+                Height::new(self.chain_id.version(), height)
+                    .wrap_err_with(|| format!(
+                        "Failed to create Height with revision number `{}` and revision height `{height}`", 
+                        self.chain_id.version()
+                    ))
+            }).transpose()?;
+
+            let consensus_state = if let Some(query_height) = query_height {
+                chain
+                    .query_consensus_state(&self.client_id, &consensus_height, &query_height)
+                    .await
+                    .wrap_err_with(|| {
+                        format!(
+                            "failed to query consensus state at height `{}` for client `{}`",
+                            consensus_height, self.client_id
+                        )
+                    })?
+            } else {
+                chain
+                    .query_consensus_state_with_latest_height(&self.client_id, &consensus_height)
+                    .await
+                    .wrap_err_with(|| {
+                        format!(
+                            "failed to query latest consensus state at height `{}` for client `{}`",
+                            consensus_height, self.client_id
+                        )
+                    })?
+            };
+
+            Ok(Output::success(consensus_state))
+        } else {
+            let heights = <CosmosChain as CanQueryConsensusStateHeights<CosmosChain>>::query_consensus_state_heights(&chain, &self.client_id).await
```

Comment:
@soareschen Any idea why type inference sometimes fails to infer the appropriate type whereas it works in other cases, eg. for `query_consensus_state_with_latest_height` above?

Reply:
I guess it might have to do with type inference of the returned value. If Rust can infer the return type based on how it is used later, and if there is a unique `Counterparty` type that result in the return type, then Rust will use that counterparty type.

In general, a more reliable way to avoid using the fully qualified syntax is to write code generically all the way up. In abstract code, Rust do not know the concrete type of `Chain` other than that it implements a trait like `CanQueryConsensusState<Counterparty>`. So when we call a method like `.query_consensus_state()`, Rust will pick the only type it knows that satisfies the constraint, which is `Counterparty`. That's why this problem only arises when we deal directly with concrete types like `CosmosChain`.

Reply:
We'll fix this once we add Sovereign support to the CLIs